### ISSUE TITLE
Method C_DecryptInit returned CKR_OBJECT_HANDLE_INVALID

### ISSUE THREAD

**Issue by mcabriel** [Pkcs11Interop/Pkcs11Interop#135]
Hi,

I'm using your this library and doing the Encrypt using mechanism `CKM.CKM_DES3_CBC`.
Then stored that encrypted data. 
Afterwards, I decrypted that data on different instance and encountered this issue `Method C_DecryptInit returned CKR_OBJECT_HANDLE_INVALID`.

I have read some explanations regarding this error in the pdf document that "B1 attempts to perform some operation.....".

Although in doing the test under PKCS11Interop `_01_EncryptAndDecryptSinglePartTest`, i tweaked it by doing the same encryption and decryption separately. It works properly  using the pkcs11-mock dll with the correct decrypted data result. But when using the softhsm2 dll, that's when I encountered this problem.

Any workarounds about this?

Below are the test codes
**Uses softhsm2 dll instead of the mock dll

```csharp
       [Test()]
        public void _01_EncryptAndDecryptSinglePartTest()
        {

            var cardNo = "ABDCefgh4141";

            var resultEncrypted = Encrypt(cardNo);
            var resultDecryptedText = Decrypt(resultEncrypted);
        }

        public byte[] Encrypt(string text)
        {
            Pkcs11InteropFactories factories = new Pkcs11InteropFactories();

            using (IPkcs11Library pkcs11Library = factories.Pkcs11LibraryFactory.LoadPkcs11Library(factories, Settings.Pkcs11LibraryPath, AppType.MultiThreaded))
            {
                ISlot slot = Helpers.GetUsableSlot(pkcs11Library);

                using (ISession session = slot.OpenSession(SessionType.ReadWrite))
                {
                    session.Login(CKU.CKU_USER, Settings.NormalUserPin);

                    // Generate symetric key
                    IObjectHandle generatedKey = Helpers.GenerateKey(session);
                    this.GeneratedKey = generatedKey;

                    // Generate initialization vector
                    byte[] iv = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };

                    // Specify encryption mechanism with initialization vector as parameter
                    IMechanism mechanism = session.Factories.MechanismFactory.Create(CKM.CKM_DES3_CBC, iv);

                    byte[] sourceData = ConvertUtils.Utf8StringToBytes(text);
                    PadToMultipleOf(ref sourceData, 16);

                    // Encrypt data
                    byte[] encryptedData = session.Encrypt(mechanism, generatedKey, sourceData);

                    byte[] decryptedData = session.Decrypt(mechanism, generatedKey, encryptedData);
                    var strText = ConvertUtils.BytesToUtf8String(decryptedData);

                    return encryptedData;
                }
            }
        }

        public string Decrypt(byte[] encryptedData)
        {
            Pkcs11InteropFactories factories = new Pkcs11InteropFactories();

            using (IPkcs11Library pkcs11Library = factories.Pkcs11LibraryFactory.LoadPkcs11Library(factories, Settings.Pkcs11LibraryPath, AppType.MultiThreaded))
            {
                ISlot slot = Helpers.GetUsableSlot(pkcs11Library);

                using (ISession session = slot.OpenSession(SessionType.ReadWrite))
                {
                    session.Login(CKU.CKU_USER, Settings.NormalUserPin);

                    // Generate symetric key
                    //IObjectHandle generatedKey = PKCS11InteropHelpers.GenerateKey(session);
                    IObjectHandle generatedKey = (IObjectHandle)this.GeneratedKey;

                    // Generate initialization vector
                    byte[] iv = new byte[] { 0, 0, 0, 0, 0, 0, 0, 0 };

                    // Specify encryption mechanism with initialization vector as parameter
                    IMechanism mechanism = session.Factories.MechanismFactory.Create(CKM.CKM_DES3_CBC, iv);

                    // Decrypt data
                    byte[] decryptedData = session.Decrypt(mechanism, generatedKey, encryptedData);

                    return ConvertUtils.BytesToUtf8String(decryptedData);
                }
            }
        }
```

**Comment by jariq** [comment]
Hello @mcabriel,

you're most likely getting `CKR_OBJECT_HANDLE_INVALID` error because object handles are not guaranteed to be the same between library loads.

It is highly recommended that before you start using Pkcs11Interop you get familiar at least with *"Chapter 2 - Scope"*, *"Chapter 6 - General overview"* and *"Chapter 10 - Objects"* of [PKCS#11 v2.20](https://github.com/Pkcs11Interop/PKCS11-SPECS/tree/master/v2.20) specification (or equivalent chapters of any previous or subsequent specification version).

Please also note that PKCS11-MOCK is not a real cryptographic module but just a dummy mock object designed specifically for unit testing of Pkcs11Interop library.

**Comment by mcabriel** [comment]
Hi @jariq,

Thanks for the quick response. I went through the pdf document and this makes sense to me.
On the same application, on the Encrypt() -> it closed the session that destroys the object handle.

Is it possible to copy or create the object handle during Encrypt() level "C_CopyObject" or "C_CreateObject" and find it "C_FindObjects" or create it on the Decrypt() level?

Although you've mentioned that the object handles are not guaranteed to be the same between library loads, does this mean that this library load should be 1 time (on all related operations) and always open?

**Comment by jariq** [comment]
You might want to read "Chapter 10 - Objects" of [PKCS#11 v2.20](https://github.com/Pkcs11Interop/PKCS11-SPECS/tree/master/v2.20) specification again because you seem to be misunderstanding what object handle is and how it should be used.

In general there are objects (e.g. keys) stored on your token. If you want to work with existing object (e.g. key) you need to acquire its object handle. Object handle is in many ways akin to a file handle that you use when you work with files. To acquire an object handle to an existing object you need to find it with `C_FindObject*` functions.

I'm just guessing here but it seems to me that you are blindly copying `_01_EncryptAndDecryptSinglePartTest()`[https://github.com/Pkcs11Interop/Pkcs11Interop/blob/5.0.0/src/Pkcs11Interop.Tests/HighLevelAPI/_19_EncryptAndDecryptTest.cs#L42] without understanding what it does. It creates temporary key, demonstrates encryption and decryption with that key and then deletes the key. This is most likely not what you would do in a real world application.

**Comment by mcabriel** [comment]
Thanks @jariq,

I'll go through it again.

By the way, I've been doing general test for `C_CreateObject()` and `C_CopyObject()` separately using SoftHSM in windows, and it seems that the platform are not supported for `C_CopyObject()`, is that right?

**Comment by jariq** [comment]
I've seen you mentioned this problem also in https://github.com/opendnssec/SoftHSMv2/issues/514 and it really looks like SoftHSM returns `CKR_ARGUMENTS_BAD` when `C_CopyObject` is called without [`pTemplate` parameter specified](https://github.com/opendnssec/SoftHSMv2/blob/2.5.0/src/lib/SoftHSM.cpp#L1469).

I'm not sure why SoftHSM is doing that but you can workaround this limitation by changing the line

```csharp
IObjectHandle copiedObjectHandle = session.CopyObject(objectHandle, null);
```

to 

```csharp
IObjectHandle copiedObjectHandle = session.CopyObject(objectHandle, new List<IObjectAttribute>());
```

**Comment by mcabriel** [comment]
Thanks @jariq, appreciated!

**Comment by mcabriel** [comment]
Hi @jariq,

I've read several Chapters of [PKCS#11 v2.20](https://github.com/Pkcs11Interop/PKCS11-SPECS/tree/master/v2.20).

Based on the PKCS11Interop for c#, the `session.Encrypt` method accepts 3 parameters, namely, mechanism, generatedKey, and sourceData. I've checked your library and all methods `C_Encrpyt`, `C_EncryptInit`, `C_EcryptFinal` are similar with all of the parameters written in the document.

My goal is to encrypt the sourceData using DES3 CBC mechanism, alone. And the encrypted data will be saved to an external Database table or the likes.
This encrypted data will be decrypted by a separate session on a different instance.

As per `session.Encrypt` under the library, it accepts the generatedKey (which is the secret key that  was generated). I am aware that DES3 CBC can only be Encrypt/Decrypt and Wrap/Unwrap. And not possible to use GeneratedKeyPair that utilizes publicKey for encryption and privateKey for decryption.

My question is, based on the PKCS11Interop library for C#, is there a workaround to Encrypt a sourceData using DES3 CBC mechanism, and decrypt it on a different level/instance? <<--Session that was used for encryption and decryption are different.

**Comment by mcabriel** [comment]
Hi @jariq,

Thank you for your statement regarding the "temporary key". I have found out that the reason behind the `CKR_OBJECT_HANDLE_INVALID` in `C_DecryptInit` is because of this "temporary" key. `C_FindObjects` will not work as well since it cant in another instance. 

Based on the PKCS#11 specs pdf document, `CKR_OBJECT_HANDLE_INVALID` is not part of the return value(s) of `C_DecryptInit` which makes it more confusing.

Also, the different library loads will not be a problem as well.

I generated/exported a `CKK_DES3` secret key separately, (which actually I'm wondering why "the" `Helpers.GenerateKey` under in `Helpers` class in test projects are not creating one when I checked the SoftHSM thru a PKCS Admin GUI <<- this really helps me saw the problem).

We can close the issue now.
